mwdb/core/search/fields.py

```diff
@@ -26,348 +28,360 @@
-    """
-    statement = re.sub(r"\\(?![%_])", r"\\\\", statement)
-    return statement
+class BaseField:
+    accepts_subpath = False
 
+    def _get_condition(self, value: Item, path_selector: PathSelector) -> Any:
+        raise NotImplementedError
 
-class BaseField:
-    accepts_range = False
-    accepts_subquery = False
-    accepts_subfields = False
-    accepts_wildcards = False
+    def get_condition(self, value: Item, path_selector: PathSelector) -> Any:
+        if not self.accepts_subpath and len(path_selector) > 1:
+            raise FieldNotQueryableException("Subfields are not allowed for this field")
+        return self._get_condition(value, path_selector)
 
+
+class ColumnField(BaseField):
     def __init__(self, column):
         self.column = column
 
     @property
-    def field_type(self) -> Type[Object]:
+    def column_type(self) -> Type[Object]:
         return self.column.class_
 
-    def _get_condition(
-        self, expression: Expression, subfields: List[Tuple[str, int]]
-    ) -> Any:
-        raise NotImplementedError
-
-    def get_condition(
-        self, expression: Expression, subfields: List[Tuple[str, int]]
-    ) -> Any:
-        if not self.accepts_subfields and len(subfields) > 1:
-            raise FieldNotQueryableException(
-                f"Field doesn't have subfields: "
-                f"{'.'.join([field for field, _ in subfields[1:]])}"
-            )
-        if (
-            not self.accepts_wildcards
-            and isinstance(expression, Term)
-            and expression.has_wildcard()
-        ):
-            raise UnsupportedGrammarException(
-                "Wildcards are not allowed for this field"
-            )
-        return self._get_condition(expression, subfields)
-
-
-class StringField(BaseField):
-    accepts_wildcards = True
-
-    def _get_condition(
-        self, expression: Expression, subfields: List[Tuple[str, int]]
-    ) -> Any:
-        value = get_term_value(expression)
-        if expression.has_wildcard():
-            value = add_escaping_for_like_statement(value)
-            return self.column.like(value)
-        else:
-            return self.column == value
 
+class StringField(ColumnField):
+    def _get_condition(self, value: Item, path_selector: PathSelector) -> Any:
+        string_value = string_from_node(value, escaped=True)
+        return string_equals(self.column, string_value)
 
-class SizeField(BaseField):
-    accepts_range = True
-    accepts_wildcards = True
 
-    def _get_condition(
-        self, expression: Expression, subfields: List[Tuple[str, int]]
-    ) -> Any:
+class SizeField(ColumnField):
+    def _get_condition(self, value: Item, path_selector: PathSelector) -> Any:
         units = {"B": 1, "KB": 1024, "MB": 1024**2, "GB": 1024**3}
 
-        def parse_size(size):
-            if size.isdigit():
-                return size
+        def parse_size(size) -> int:
+            if re.match(r"^\d+$", size) is not None:
```

Comment:
what was wrong with isdigit? :thinking:

Reply:
isdigit accepts Unicode characters that are not actual digits like `६`
```
>>> "६".isdigit()
True
```
it caused ISE 500 instead of 400 Bad Request